pkg/controllers/image_controller.go

```diff
@@ -0,0 +1,472 @@
+			}
+		}()
+	}
+
+	wg.Wait()
+	return nil
+}
+
+func (r *ImageReconciler) processNextWorkItem(ctx context.Context, log logr.Logger) bool {
+	item, shutdown := r.queue.Get()
+	if shutdown {
+		return false
+	}
+	defer r.queue.Done(item)
+
+	id := item.(string)
+	log = log.WithValues("imageId", id)
+	ctx = logr.NewContext(ctx, log)
+
+	if err := r.reconcileImage(ctx, id); err != nil {
+		log.Error(err, "failed to reconcile image")
+		r.queue.AddRateLimited(item)
+		return true
+	}
+
+	r.queue.Forget(item)
+	return true
+}
+
+const (
+	ImageFinalizer = "image"
+)
+
+func (r *ImageReconciler) deleteImage(ctx context.Context, log logr.Logger, ioCtx *rados.IOContext, image *api.Image) error {
+	if !slices.Contains(image.Finalizers, ImageFinalizer) {
+		log.V(1).Info("image has no finalizer: done")
+		return nil
+	}
+
+	if err := librbd.RemoveImage(ioCtx, ImageIDToRBDID(image.ID)); err != nil && !errors.Is(err, librbd.ErrNotFound) {
+		return fmt.Errorf("failed to remove rbd image: %w", err)
+	}
+	log.V(2).Info("Rbd image deleted")
+
+	image.Finalizers = utils.DeleteSliceElement(image.Finalizers, ImageFinalizer)
+	if _, err := r.images.Update(ctx, image); store.IgnoreErrNotFound(err) != nil {
+		return fmt.Errorf("failed to update image metadata: %w", err)
+	}
+	log.V(2).Info("Removed Finalizers")
+
+	return nil
+}
+
+type fetchAuthResponse struct {
+	Key string `json:"key"`
+}
+
+func (r *ImageReconciler) fetchAuth(ctx context.Context, log logr.Logger) (string, string, error) {
+	cmd1, err := json.Marshal(map[string]string{
+		"prefix": "auth get-key",
+		"entity": r.client,
+		"format": "json",
+	})
+	if err != nil {
+		return "", "", fmt.Errorf("unable to marshal command: %w", err)
+	}
+
+	log.V(3).Info("Try to fetch client", "name", r.client)
+	data, _, err := r.conn.MonCommand(cmd1)
+	if err != nil {
+		return "", "", fmt.Errorf("failed to execute mon command: %w", err)
+	}
+
+	response := fetchAuthResponse{}
+	if err := json.Unmarshal(data, &response); err != nil {
+		return "", "", fmt.Errorf("unable to unmarshal response: %w", err)
+	}
+
+	return strings.TrimPrefix(r.client, "client."), response.Key, nil
+}
+
+func (r *ImageReconciler) reconcileSnapshot(ctx context.Context, log logr.Logger, img *api.Image) error {
+	if img.Spec.Image == "" || img.Spec.SnapshotRef != nil {
+		return nil
+	}
+
+	resolvedImg, err := r.registry.Resolve(ctx, img.Spec.Image)
+	if err != nil {
+		return fmt.Errorf("failed to resolve image ref in registry: %w", err)
+	}
+
+	snapshotDigest := resolvedImg.Descriptor().Digest.String()
+
+	//TODO select later by label
+	snap, err := r.snapshots.Get(ctx, snapshotDigest)
+	if err != nil {
+
+		switch {
+
+		case errors.Is(err, store.ErrNotFound):
+			snap, err = r.snapshots.Create(ctx, &api.Snapshot{
+				Metadata: api.Metadata{
+					ID: snapshotDigest,
+					Labels: map[string]string{
+						imageDigestLabel: snapshotDigest,
+					},
+				},
+				Source: api.SnapshotSource{
+					OnmetalImage: img.Spec.Image,
```

Comment:
The source should point to the resolved snapshot, not again to the image leading it to be potentially resolved to another digest.